File: src/openrtm_aist_python/OpenRTM_aist/Task.py
#!/usr/bin/env python
# -*- coding: euc-jp -*-

##
# @file Task.py
# @brief Task class
# @date $Date$
# @author Noriaki Ando <n-ando@aist.go.jp> and Shinji Kurihara
#
# Copyright (C) 2008
#     Task-intelligence Research Group,
#     Intelligent Systems Research Institute,
#     National Institute of
#         Advanced Industrial Science and Technology (AIST), Japan
#     All rights reserved.
#

import threading

class Task:
  def __init__(self):
    self._count = 0
    self._thread = None
    return

    
  def __del__(self):
    self._count = 0
    #if self._thread:
    #  if self._thread.isAlive():
    #    self._thread.join()
    self._thread = None
    return


  def open(self, args = None):
    return 0


  def close(self, flags = 0):
    return 0


  def svc(self):
    return 0


  def activate(self):
    if self._count == 0:
      self._thread = threading.Thread(target=self.svc_run)
      self._count += 1
      self._thread.start()
    return


  def wait(self):
    if self._count > 0:
      self._thread.join()
    return


  def suspend(self):
    return 0


  def resume(self):
    return 0


  def reset(self):
    self._count = 0
    return


  def finalize(self):
    self.reset()
    return

  def svc_run(self):
    self.svc()
    self.finalize()
    return

File: src/openrtm_aist/src/lib/coil/win32/coil/Task.cpp
// -*- C++ -*-
/*!
 * @file Task_win32.cpp
 * @brief Task class
 * @date $Date$
 * @author Noriaki Ando <n-ando@aist.go.jp>
 *
 * Copyright (C) 2008
 *     Task-intelligence Research Group,
 *     Intelligent Systems Research Institute,
 *     National Institute of
 *         Advanced Industrial Science and Technology (AIST), Japan
 *     All rights reserved.
 *
 * $Id$
 *
 */

#include <coil/Task.h>

namespace coil
{

  /*!
   * @if jp
   * @brief コンストラクタ
   * @else
   * @brief Constructor
   * @endif
   */
  Task::Task()
    : m_count(0)
  {
  }

  /*!
   * @if jp
   * @brief デストラクタ
   * @else
   * @brief Destructor
   * @endif
   */
  Task::~Task()
  {
    m_count = 0;
  }

  /*!
   * @if jp
   * @brief タスクオープン
   * @else
   * @brief Task open
   * @endif
   */
  int Task::open(void* args)
  {
    return 0;
  }

  /*!
   * @if jp
   * @brief タスククローズ
   * @else
   * @brief Task close
   * @endif
   */
  int Task::close(unsigned long flags)
  {
    return 0;
  }

  /*!
   * @if jp
   * @brief スレッドを実行する
   * @else
   * @brief Execute thread
   * @endif
   */
  int Task::svc()
  {
    return 0;
  }

  /*!
   * @if jp
   * @brief スレッドを生成する
   * @else
   * @brief Create a thread
   * @endif
   */
  void Task::activate()
  {
    if (m_count == 0)
      {
		  m_thread =
			  (HANDLE)::_beginthreadex(NULL, // security
			  0, //stuck size
			  Task::svc_run, // func
			  (void*)this, // argument
			  0, // flag (0 or CREATE_SUSPEND)
			  NULL); //thread descripter
		  ++m_count;
      };
  }

  /*!
   * @if jp
   * @brief スレッド終了を待つ
   * @else
   * @brief Waiting for the thread terminate
   * @endif
   */
  int Task::wait(void)
  {
    if (m_count > 0)
      {
        DWORD retval;
	retval = ::WaitForSingleObject(m_thread, INFINITE);
      }
    return 0;
  }

  /*!
   * @if jp
   * @brief タスク実行を中断する
   * @else
   * @brief Suspending the task
   * @endif
   */
  int Task::suspend(void)
  {
    return 0;
  }

  /*!
   * @if jp
   * @brief 中断されているタスクを再開する
   * @else
   * @brief Resuming the suspended task
   * @endif
   */
  int Task::resume(void)
  {
    return 0;
  }

  /*!
   * @if jp
   * @brief タスク数リセット
   * @else
   * @brief Reset of task count
   * @endif
   */
  void Task::reset()
  {
    m_count = 0;
  }

  /*!
   * @if jp
   * @brief タスク実行を終了する
   * @else
   * @brief Finalizing the task
   * @endif
   */
  void Task::finalize()
  {
    reset();
  }
 
  /*!
   * @if jp
   * @brief スレッド実行を開始する
   * @else
   * @brief Start thread Execution
   * @endif
   */
  unsigned int WINAPI Task::svc_run(void* args)
  {
    Task* t = (coil::Task*)args;
    int status;
    status = t->svc();
    t->finalize();
    return 0;
  }
};


